solr/core/src/java/org/apache/solr/util/tracing/SimplePropagator.java

```diff
@@ -0,0 +1,304 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.util.tracing;
+
+import io.opentracing.Scope;
+import io.opentracing.ScopeManager;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.Tracer.SpanBuilder;
+import io.opentracing.propagation.Format;
+import io.opentracing.propagation.TextMap;
+import io.opentracing.tag.Tag;
+import io.opentracing.util.GlobalTracer;
+import io.opentracing.util.ThreadLocalScopeManager;
+import java.lang.invoke.MethodHandles;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SimplePropagator {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private static final String TRACE_HOST_NAME =
+      System.getProperty("trace_host_name", System.getProperty("host"));
+
+  static final String TRACE_ID = System.getProperty("TRACE_ID", "X-Trace-Id");
+
+  private static final AtomicLong traceCounter = new AtomicLong(0);
+
+  public static void load() {
+    GlobalTracer.registerIfAbsent(
+        () -> {
+          log.info("Always-on trace id generation enabled.");
+          return new SimplePropagatorTracer();
+        });
+  }
+
+  /**
+   * Tracer that only aims to do simple header propagation, tailored to how Solr works.
+   *
+   * <p>Heavily inspired from JaegerTracer, NoopTracer
+   */
+  static class SimplePropagatorTracer implements Tracer {
+
+    private final ScopeManager scopeManager = new ThreadLocalScopeManager();
+
+    @Override
+    public ScopeManager scopeManager() {
+      return scopeManager;
+    }
+
+    @Override
+    public Span activeSpan() {
+      return scopeManager.activeSpan();
+    }
+
+    @Override
+    public Scope activateSpan(Span span) {
+      return scopeManager.activate(span);
+    }
+
+    @Override
+    public SpanBuilder buildSpan(String operationName) {
+      return new SimplePropagatorSpanBuilder(scopeManager);
+    }
+
+    @Override
+    public void close() {}
+
+    @Override
+    public <C> void inject(SpanContext spanContext, Format<C> format, C carrier) {
+      if (!format.equals(Format.Builtin.HTTP_HEADERS)) {
+        throw new IllegalArgumentException("unsupported format " + format);
+      }
```

Comment:
Is this constraint necessary?  What if we inject & extract to transfer to the Overseer?  (I have done this at work)

Reply:
In practice at the moment, the format will *always* be HTTP since we control where we inject.  But if we get to add the Overseer as you have done in main branch, then it will mysteriously not propagate, and then maybe we'll "fix" it by removing these 3 lines.  I don't think need to care about what the format is; just try to inject best-effort.

Reply:
> I don't think need to care about what the format is; just try to inject best-effort.
I agree. the `inject` is easy enough, you just return early. the `extract` is now a `return NoopSpan.INSTANCE.context();` and this I'm not sure will work correctly.
not having some basic test coverage for this setup worries me a bit.